src/Emogrifier/CssConcatenator.php

```diff
@@ -0,0 +1,122 @@
+     *
+     * @var stdClass[]
+     */
+    private $mediaRules = [];
+
+    /**
+     * Appends a declaration block to the CSS.
+     *
+     * @param string[] $selectors Array of selectors for the rule, e.g. ["ul", "ol", "p:first-child"].
+     * @param string $declarationsBlock The property declarations, e.g. "margin-top: 0.5em; padding: 0".
+     * @param string $media The media query for the rule, e.g. "@media screen and (max-width:639px)",
+     *                      or an empty string if none.
+     */
+    public function append(array $selectors, $declarationsBlock, $media = '')
+    {
+        $lastMediaRule = end($this->mediaRules);
+        if ($lastMediaRule !== false && $media === $lastMediaRule->media) {
+            $mediaRule = $lastMediaRule;
+        } else {
+            $mediaRule = (object)[
+                'media' => $media,
+                'ruleBlocks' => [],
+            ];
+            $this->mediaRules[] = $mediaRule;
+        }
+
+        $lastRuleBlock = end($mediaRule->ruleBlocks);
+        $selectorsAsKeys = array_flip($selectors);
+        if ($lastRuleBlock !== false && $declarationsBlock === $lastRuleBlock->declarationsBlock) {
+            $lastRuleBlock->selectorsAsKeys += $selectorsAsKeys;
+        } elseif ($lastRuleBlock !== false
+            && $this->hasEquivalentSelectors($selectorsAsKeys, $lastRuleBlock->selectorsAsKeys)
+        ) {
+            $lastRuleBlock->declarationsBlock
+                = rtrim(rtrim($lastRuleBlock->declarationsBlock), ';') . ';' . $declarationsBlock;
+        } else {
+            $mediaRule->ruleBlocks[] = (object)compact('selectorsAsKeys', 'declarationsBlock');
+        }
+    }
+
+    /**
+     * @return string
+     */
+    public function getCss()
+    {
+        $css = '';
+        foreach ($this->mediaRules as $mediaRule) {
+            if ($mediaRule->media !== '') {
+                $css .= $mediaRule->media . '{';
+            }
+            foreach ($mediaRule->ruleBlocks as $ruleBlock) {
+                $css .= implode(',', array_keys($ruleBlock->selectorsAsKeys))
+                    . '{' . $ruleBlock->declarationsBlock . '}';
+            }
+            if ($mediaRule->media !== '') {
+                $css .= '}';
+            }
+        }
+        return $css;
+    }
+
+    /**
+     * Tests if two sets of selectors are equivalent (i.e. the same selectors, possibly in a different order).
+     *
+     * @param int[] $selectorsAsKeys1 Array in which the selectors are the keys, and the values are of no significance.
+     * @param int[] $selectorsAsKeys2 Another such array.
+     *
+     * @return bool
+     */
+    private static function hasEquivalentSelectors(array $selectorsAsKeys1, array $selectorsAsKeys2)
+    {
+        return count($selectorsAsKeys1) === count($selectorsAsKeys2)
+            && count($selectorsAsKeys1) === count($selectorsAsKeys1 + $selectorsAsKeys2);
```

Comment:
Will `count($selectorsAsKeys1 + $selectorsAsKeys2)` be the same as `count($selectorsAsKeys1) + count($selectorsAsKeys2)`? If so, we can put the two latter counts into local variables and then work on those, allowing us to have only two count calls.

Reply:
`count(A+B)` gives the number of unique keys in the union of A and B, and will be lower than `count(A) + count(B)` whenever there are common keys.